plugins/spi/SPIOutput.cpp

```diff
@@ -871,7 +884,111 @@ uint8_t SPIOutput::CalculateAPA102PixelBrightness(uint8_t brightness) {
+                                       buffer.Size() - first_slot);
+
+  for (unsigned int i = 0; i < length / WS2812B_SLOTS_PER_PIXEL; i++) {
+    // Convert RGB to GRB
+    unsigned int offset = first_slot + i * WS2812B_SLOTS_PER_PIXEL;
+    uint8_t r = buffer.Get(offset);
+    uint8_t g = buffer.Get(offset + 1);
+    uint8_t b = buffer.Get(offset + 2);
+    uint8_t low = 0, mid = 0, high = 0;
+
+    WS2812bByteMapper(g, low, mid, high);
+    output[i * WS2812B_SPI_BYTES_PER_PIXEL] = low;
+    output[i * WS2812B_SPI_BYTES_PER_PIXEL + 1] = mid;
+    output[i * WS2812B_SPI_BYTES_PER_PIXEL + 2] = high;
+
+    WS2812bByteMapper(r, low, mid, high);
+    output[i * WS2812B_SPI_BYTES_PER_PIXEL + 3] = low;
+    output[i * WS2812B_SPI_BYTES_PER_PIXEL + 4] = mid;
+    output[i * WS2812B_SPI_BYTES_PER_PIXEL + 5] = high;
+
+    WS2812bByteMapper(b, low, mid, high);
+    output[i * WS2812B_SPI_BYTES_PER_PIXEL + 6] = low;
+    output[i * WS2812B_SPI_BYTES_PER_PIXEL + 7] = mid;
+    output[i * WS2812B_SPI_BYTES_PER_PIXEL + 8] = high;
+  }
+
+  // write output back...
+  m_backend->Commit(m_output_number);  
+}
+
+void SPIOutput::CombinedWS2812bControl(const DmxBuffer &buffer) {
+  const unsigned int first_slot = m_start_address - 1;  // 0 offset
+  if (buffer.Size() - first_slot < WS2812B_SLOTS_PER_PIXEL) {
+    OLA_INFO << "Insufficient DMX data, required " << WS2812B_SLOTS_PER_PIXEL
+             << ", got " << buffer.Size() - first_slot;
+    return;
+  }
+
+  // We always check out the entire string length, even if we only have data
+  // for part of it
+  const unsigned int output_length = m_pixel_count * WS2812B_SPI_BYTES_PER_PIXEL;
+  uint8_t *output = m_backend->Checkout(m_output_number, output_length);
+  if (!output) {
+    OLA_INFO << "Unable to create output buffer of required length: " << output_length;
+    return;
+  }
+
+  // Grab RGB data for conversion to GBR
+  uint8_t r = buffer.Get(first_slot);
+  uint8_t g = buffer.Get(first_slot + 1);
+  uint8_t b = buffer.Get(first_slot + 2);
+  uint8_t low = 0, mid = 0, high = 0;
 
+  // create Pixel Data
+  uint8_t pixel_data[WS2812B_SPI_BYTES_PER_PIXEL];
+
+  WS2812bByteMapper(g, low, mid, high);
+  pixel_data[0] = low;
+  pixel_data[1] = mid;
+  pixel_data[2] = high;
+
+  WS2812bByteMapper(r, low, mid, high);
+  pixel_data[3] = low;
+  pixel_data[4] = mid;
+  pixel_data[5] = high;
+
+  WS2812bByteMapper(b, low, mid, high);
+  pixel_data[6] = low;
+  pixel_data[7] = mid;
+  pixel_data[8] = high;
+
+  // set all pixel to same value
+  for (uint16_t i = 0; i < m_pixel_count; i++) {
+    memcpy(&output[i * WS2812B_SPI_BYTES_PER_PIXEL], pixel_data,
+           WS2812B_SPI_BYTES_PER_PIXEL);
+  }
+
+  // write output back...
+  m_backend->Commit(m_output_number);
+}
+
+void SPIOutput::WS2812bByteMapper(uint8_t input, uint8_t &low, uint8_t &mid, uint8_t &high)
+{
+  low = 0x24 | ((input & 0x1) << 1) | ((input & 0x2) << 3) | ((input & 0x4) << 5);
```

Comment:
Can we have some comment or explanation of these numbers/what's going on generally with the way it works please. Or a link to a page with some detail.